tianshou/data/buffer.py

```diff
@@ -241,117 +174,109 @@ def add(
         **kwargs: Any,
-    ) -> None:
-        """Add a batch of data into replay buffer."""
+    ) -> Tuple[int, float]:
+        """Add a batch of data into replay buffer.
+
+        Return (episode_length, episode_reward) if one episode is terminated,
+        otherwise return (0, 0.0).
+        """
         assert isinstance(
             info, (dict, Batch)
         ), "You should return a dict in the last argument of env.step()."
-        if self._last_obs:
+        if self._save_only_last_obs:
             obs = obs[-1]
         self._add_to_buffer("obs", obs)
         self._add_to_buffer("act", act)
-        # make sure the reward is a float instead of an int
-        self._add_to_buffer("rew", rew * 1.0)  # type: ignore
-        self._add_to_buffer("done", done)
-        if self._save_s_:
+        # make sure the data type of reward is float instead of int
+        rew = rew * 1.0  # type: ignore
+        self._add_to_buffer("rew", rew)
+        self._add_to_buffer("done", bool(done))  # done should be a bool scalar
+        if self._save_obs_next:
             if obs_next is None:
                 obs_next = Batch()
-            elif self._last_obs:
+            elif self._save_only_last_obs:
                 obs_next = obs_next[-1]
             self._add_to_buffer("obs_next", obs_next)
         self._add_to_buffer("info", info)
         self._add_to_buffer("policy", policy)
 
-        # maintain available index for frame-stack sampling
-        if self._avail:
-            # update current frame
-            avail = sum(self.done[i] for i in range(
-                self._index - self.stack_num + 1, self._index)) == 0
-            if self._size < self.stack_num - 1:
-                avail = False
-            if avail and self._index not in self._avail_index:
-                self._avail_index.append(self._index)
-            elif not avail and self._index in self._avail_index:
-                self._avail_index.remove(self._index)
-            # remove the later available frame because of broken storage
-            t = (self._index + self.stack_num - 1) % self._maxsize
-            if t in self._avail_index:
-                self._avail_index.remove(t)
-
-        if self._maxsize > 0:
-            self._size = min(self._size + 1, self._maxsize)
-            self._index = (self._index + 1) % self._maxsize
+        self._episode_reward += rew
+        self._episode_length += 1
+
+        if self.maxsize > 0:
+            self._size = min(self._size + 1, self.maxsize)
+            self._index = (self._index + 1) % self.maxsize
+        else:  # TODO: remove this after deleting ListReplayBuffer
+            self._size = self._index = self._size + 1
+
+        if done:
+            result = (self._episode_length, self._episode_reward)
+            self._episode_length, self._episode_reward = 0, 0.0
+            return result
         else:
-            self._size = self._index = self._index + 1
+            return (0, 0.0)
```

Comment:
can 'add' in Replaybuffer return np.array just to be consistent with cachedbuffers?

Reply:
What's the dimension you want, 0-dim or 1-dim?

Reply:
Both seems okay, 0-dim is clearer but 1-dim is more consistent